File: go/dynamicprovider/signing.go
package dynamicprovider

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
)

func SignRequest(method, path, timestamp string, body, secret []byte) string {
	digest := sha256.Sum256(body)
	payload := strings.Join([]string{
		"cortex-dynamic-provider-request",
		ProtocolVersion,
		strings.ToUpper(strings.TrimSpace(method)),
		strings.TrimSpace(path),
		strings.TrimSpace(timestamp),
		hex.EncodeToString(digest[:]),
	}, "\n")
	return computeHMAC(secret, payload)
}

func SignResponse(method, path, timestamp string, statusCode int, body, secret []byte) string {
	digest := sha256.Sum256(body)
	payload := strings.Join([]string{
		"cortex-dynamic-provider-response",
		ProtocolVersion,
		strings.ToUpper(strings.TrimSpace(method)),
		strings.TrimSpace(path),
		fmt.Sprintf("%d", statusCode),
		strings.TrimSpace(timestamp),
		hex.EncodeToString(digest[:]),
	}, "\n")
	return computeHMAC(secret, payload)
}

func VerifyRequestSignature(method, path, timestamp string, body []byte, signature string, secret []byte) bool {
	expected := SignRequest(method, path, timestamp, body, secret)
	return hmac.Equal([]byte(signature), []byte(expected))
}

func computeHMAC(secret []byte, payload string) string {
	h := hmac.New(sha256.New, secret)
	_, _ = h.Write([]byte(payload))
	return hex.EncodeToString(h.Sum(nil))
}

File: python/governance_sdk/signing.py
from __future__ import annotations

import hashlib
import hmac
import os

from .models import PROTOCOL_VERSION


def secret_from_env() -> bytes:
    secret = os.getenv("GOVERNANCE_DYNAMIC_PROVIDER_SECRET") or os.getenv("GOVERNANCE_SECRET")
    if not secret:
        raise RuntimeError("set GOVERNANCE_DYNAMIC_PROVIDER_SECRET or GOVERNANCE_SECRET")
    return secret.encode("utf-8")


def sign_request(method: str, path: str, timestamp: str, body: bytes, secret: bytes) -> str:
    digest = hashlib.sha256(body).hexdigest()
    payload = "\n".join(
        [
            "cortex-dynamic-provider-request",
            PROTOCOL_VERSION,
            method.strip().upper(),
            path.strip(),
            timestamp.strip(),
            digest,
        ]
    )
    return hmac.new(secret, payload.encode("utf-8"), hashlib.sha256).hexdigest()


def sign_response(method: str, path: str, timestamp: str, status_code: int, body: bytes, secret: bytes) -> str:
    digest = hashlib.sha256(body).hexdigest()
    payload = "\n".join(
        [
            "cortex-dynamic-provider-response",
            PROTOCOL_VERSION,
            method.strip().upper(),
            path.strip(),
            str(status_code),
            timestamp.strip(),
            digest,
        ]
    )
    return hmac.new(secret, payload.encode("utf-8"), hashlib.sha256).hexdigest()


def verify_request(method: str, path: str, timestamp: str, body: bytes, signature: str, secret: bytes) -> bool:
    return hmac.compare_digest(signature, sign_request(method, path, timestamp, body, secret))
